framework/Models/PostProcessors/Validations/PhysicsGuidedCoverageMapping.py

```diff
@@ -0,0 +1,202 @@
+    return evaluation
+
+  ### utility functions
+  def _evaluate(self, datasets, **kwargs):
+    """
+      Main method to "do what you do".
+      @ In, datasets, list, list of datasets (data1,data2,etc.) to used.
+      @ In, kwargs, dict, keyword arguments
+      @ Out, outputDict, dict, dictionary containing the results {"feat"_"target"_"metric_name":value}
+    """
+    names = kwargs.get('dataobjectNames')
+    outputDict = {}
+    for feat, targ, msr in zip(self.features, self.targets, self.measurements):
+      featData = self._getDataFromDataDict(datasets, feat, names)
+      targData = self._getDataFromDataDict(datasets, targ, names)
+      msrData = self._getDataFromDataDict(datasets, msr, names)
+
+    # Values from first entry of tuple data from _getDataFromDataDict
+    yExp = np.array(featData[0])
+    yApp = np.array(targData[0])
+    yMsr = np.array(msrData[0])
+    # Sample mean as reference value
+    yExpRef = np.mean(yExp, axis=0)
+    yAppRef = np.mean(yApp)
+    # Standardization
+    yExpStd = (yExp-yExpRef)/yExpRef
+    yAppStd = (yApp-yAppRef)/yAppRef
+    if yExp.shape[1]!= yMsr.shape[1]:
+      self.raiseAnError(IOError, "Number of Measurements is not consistent with number of Experiments/Features.")
+    yMsrStd = (yMsr-yExpRef)/yExpRef
+
+    # Single Experiment response
+    if yExpStd.shape[1]==1:
+      yExpReg = yExpStd.flatten()
+    # Pseudo response of multiple Experiment responses
+    # OrthogonalMatchingPursuit from sklearn used here
+    # Possibly change to other regressors      
+    elif yExpStd.shape[1]>1:
+      regrExp = OrthogonalMatchingPursuit(fit_intercept=False).fit(yExpStd, yAppStd)
+      yExpReg = regrExp.predict(yExpStd)
+    
+    # Kernel Desnity Estimation
+    m1 = yExpReg[:]
+    m2 = yAppStd.flatten()
+    xmin = m1.min()
+    xmax = m1.max()
+    ymin = m2.min()
+    ymax = m2.max()
+
+    binKDE = 200j
+    # Grid of Experiment (X), grid of Application (Y)
+    X, Y = np.mgrid[xmin:xmax:binKDE, ymin:ymax:binKDE]
+    psts = np.vstack([X.ravel(), Y.ravel()])
+    vals = np.vstack([m1, m2])
+    # kernel
+    knl = stats.gaussian_kde(vals)
+    # Joint PDF of Experiment and Application
+    Z = np.reshape(knl(psts).T, X.shape)
+
+    if yMsrStd.shape[1]==1:
+      yMsrStd = yMsrStd.flatten()
+    # Combine measurements by multiple Experiment regression
+    elif yMsrStd.shape[1]>1:
+      yMsrStd = regrExp.predict(yMsrStd)
+    
+    # Measurement PDF with KDE
+    knlMsr = stats.gaussian_kde(yMsrStd)
+    pdfMsr = knlMsr(X[:, 0])
+
+    # yAppPred by integrating p(yexp, yapp)dyexp * p(yexp)
+    # on range [ymsr.min(), ymsr.max()]
+    pdfAppPred = np.zeros(Y.shape[1])
+    intgrPdf = 0.0 # for normalization
+    for i in range(len(Y[0, :])):
+      pAppi = 0.0
+      for j in range(len(X[:, 0])):
+        pAppi += knl.evaluate([X[j, 0], Y[0, i]]) * np.diff(X[:, 0])[0] * pdfMsr[j]
+      pdfAppPred[i] = pAppi
+      intgrPdf += pdfAppPred[i]*(Y[0, 1]-Y[0, 0])
```

Comment:
First, could you check if there is a better way to do the integration?
Second, could you check the equation? it seems to me you need to integrate over the cell (4 point in X, Y space).

Reply:
Not in use any more, removed.

Reply:
Changed to dot product.